### PR TITLE
Support for Pool1d layer for OpenCV and OpenCL targets

### PR DESCRIPTION
**Pull request by sl-sergei**
Resolves: https://github.com/opencv/opencv/issues/18643

**Merge with extra:** https://github.com/opencv/opencv_extra/pull/820

<cut/>

```
opencv_extra=test_pool1d

force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2021.1.0:20.04
build_image:Custom Win=openvino-2021.1.0
build_image:Custom Mac=openvino-2021.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/layers/pooling_layer.cpp [hunk 1]
```diff
@@ -453,14 +453,17 @@ virtual Ptr<BackendNode> initNgraph(const std::vector<Ptr<BackendWrapper> >& inp
...
+                      src.dims == 3 || src.dims == 4 || src.dims == 5, dst.dims == 3 || dst.dims == 4 || dst.dims == 5,
                       (((poolingType == ROI || poolingType == PSROI) &&
                       dst.size[0] == rois.size[0]) || src.size[0] == dst.size[0]),
                       poolingType == PSROI || src.size[1] == dst.size[1],
                       (mask.empty() || (mask.type() == src.type() && mask.size == dst.size)));
 
             PoolingInvoker p;
 
+            bool isConv1D = src.dims == 3;
+            bool isConv3D = src.dims == 5;
```
Comment (dkurt) [inline modules/dnn/src/layers/pooling_layer.cpp:465]: conv?

**Comment by sl-sergei** [comment]
@dkurt @alalek I changed code for Halide backend a little bit, it does seem correct, but do our tests cover this code?

File: modules/dnn/include/opencv2/dnn/all_layers.hpp [hunk 2]
```diff
@@ -248,7 +248,6 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
         int type;
         std::vector<size_t> kernel_size, strides;
         std::vector<size_t> pads_begin, pads_end;
-        CV_DEPRECATED_EXTERNAL Size kernel, stride, pad;
         CV_DEPRECATED_EXTERNAL int pad_l, pad_t, pad_r, pad_b;
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/all_layers.hpp:251]: Will it be complicated to remove deprecated paddings as well?

Comment (sl-sergei) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp:251]: > Will it be complicated to remove deprecated paddings as well?

Done.

File: modules/dnn/src/layers/pooling_layer.cpp [hunk 3]
```diff
@@ -214,7 +207,11 @@ class PoolingLayerImpl CV_FINAL : public PoolingLayer
             if (kernel_size.empty() || kernel_size.size() == 2)
                 return backendId == DNN_BACKEND_OPENCV ||
                        (backendId == DNN_BACKEND_HALIDE && haveHalide() &&
-                           (type == MAX || (type == AVE && !pad_t && !pad_l && !pad_b && !pad_r)));
+                       (type == MAX || (type == AVE && !pads_begin[0] && !pads_begin[1] && !pads_end[0] && !pads_end[1])));
+            if (kernel_size.size() == 1)
+                return (backendId == DNN_BACKEND_OPENCV && preferableTarget == DNN_TARGET_CPU) ||
+                       (backendId == DNN_BACKEND_OPENCV && preferableTarget == DNN_TARGET_OPENCL) ||
+                       (backendId == DNN_BACKEND_OPENCV && preferableTarget == DNN_TARGET_OPENCL_FP16);
```
Comment (alalek) [inline modules/dnn/src/layers/pooling_layer.cpp:214]: > supportBackend(int backendId)

is not supposed to check `preferableTarget` value at all.

---

It is a design bug.
Caller doesn't expect failures for `DNN_BACKEND_OPENCV`.

/cc @dkurt 

---

`return backendId == DNN_BACKEND_OPENCV;` should be enough here (replace these 3 lines).

**Comment by sl-sergei** [comment]
@dkurt Can you please review the recent changes?

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/dnn/src/layers/pooling_layer.cpp (hunk 3 above)
Comment (dkurt) [reply modules/dnn/src/layers/pooling_layer.cpp:214]: Actually, there is already `else if (backendId == DNN_BACKEND_OPENCV || backendId == DNN_BACKEND_HALIDE)` so this condition is always true

**Comment by dkurt** [comment]
Please take a look at failed test: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1440.

File: modules/dnn/src/layers/pooling_layer.cpp [hunk 4]
```diff
@@ -205,18 +198,22 @@ class PoolingLayerImpl CV_FINAL : public PoolingLayer
...
-                return backendId == DNN_BACKEND_OPENCV ||
-                       (backendId == DNN_BACKEND_HALIDE && haveHalide() &&
-                           (type == MAX || (type == AVE && !pad_t && !pad_l && !pad_b && !pad_r)));
-            else
-                return false;
+                return preferableTarget == DNN_TARGET_CPU;
+            if (kernel_size.size() == 1)
+                return (preferableTarget == DNN_TARGET_CPU ||
+                        preferableTarget == DNN_TARGET_OPENCL ||
+                        preferableTarget == DNN_TARGET_OPENCL_FP16);
```
Comment (dkurt) [inline modules/dnn/src/layers/pooling_layer.cpp:210]: Do we have other targets?

Comment (sl-sergei) [reply modules/dnn/src/layers/pooling_layer.cpp:210]: > Do we have other targets?

I just try to revert my changes to see why the tests broke. Everything was fine until I changed logic for support checks. See, for instance, https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1438

File: modules/dnn/src/layers/pooling_layer.cpp (hunk 3 above)
Comment (alalek) [reply modules/dnn/src/layers/pooling_layer.cpp:214]: It is `false` if `backendId == DNN_BACKEND_HALIDE`.

**Comment by sl-sergei** [comment]
@alalek Do you know what is wrong with Cmake warnings on Mac? 
https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1441/steps/cmake/logs/warnings%20%281%29

**Comment by alalek** [comment]
> Cmake warnings on Mac

ignore, not related to this PR

**Review by dkurt — APPROVED** [review]
:+1: Thank you!

**Pull request merged**